File: VideoCaptureInFSharp/Capture.fs
namespace VideoCaptureInFSharp

open System
open System.Drawing
open System.IO
open MonoTouch.Foundation
open MonoTouch.UIKit
open MonoTouch.AVFoundation
open MonoTouch.CoreMedia
open MonoTouch.CoreGraphics
open MonoTouch.CoreVideo
open MonoTouch.AssetsLibrary

type Recording =
    {
        VideoUrl : NSUrl
        Session : AVCaptureSession
        Writer : AVAssetWriter
        InputWriter : AVAssetWriterInput
    }

[<AutoOpen>]
module private __ =
    // SEE: http://stackoverflow.com/questions/10719770/is-there-anyway-to-use-c-sharp-implicit-operators-from-f
    let inline (!>) (x:^a) : ^b = ((^a or ^b) : (static member op_Implicit : ^a -> ^b) x) 

    let alert msg =
        let obj = new NSString() :> NSObject
        let alert = new UIAlertView("Trouble", msg, null, "OK", null)
        new NSAction(fun () -> alert.Show()) |> obj.InvokeOnMainThread

    let makeToggleButton recordToggle =
        let tb = UIButton.FromType (UIButtonType.RoundedRect)
        tb.SetTitle ("Record", UIControlState.Normal)
        let bds = UIScreen.MainScreen.Bounds
        let sz = tb.IntrinsicContentSize
        let pt = new PointF ((bds.Width - sz.Width) / 2.0f, bds.Height - sz.Height - 50.0f)
        tb.Frame <- new RectangleF (pt, sz)
        tb.TouchUpInside.Add (fun _ -> recordToggle tb)
        tb 

    let imageFromSampleBuffer (sampleBuffer : CMSampleBuffer) =
        // Get the CoreVideo image
        use pb = sampleBuffer.GetImageBuffer() :?> CVPixelBuffer
        // Lock the base address
        pb.Lock(CVOptionFlags.None) |> ignore
        let flags = CGBitmapFlags.PremultipliedFirst ||| CGBitmapFlags.ByteOrder32Little
        // Create a CGImage on the RGB colorspace from the configured parameter above
        use cs = CGColorSpace.CreateDeviceRGB()
        use context = new CGBitmapContext(pb.BaseAddress, pb.Width, pb.Height, 8, pb.BytesPerRow, cs, flags)
        use cgImage = context.ToImage()
        pb.Unlock(CVOptionFlags.None) |> ignore
        UIImage.FromImage(cgImage)

    let initializeInputWriter () =
        try
            let dictionary =
                let objects = [|AVVideo.CodecH264; new NSNumber(640); new NSNumber(480)|] : NSObject []
                let keys = [|AVVideo.CodecKey; AVVideo.WidthKey; AVVideo.HeightKey|] : NSObject []
                NSDictionary.FromObjectsAndKeys(objects, keys)

            Some(new AVAssetWriterInput(!> AVMediaType.Video, dictionary, ExpectsMediaDataInRealTime = true))
        with
            | e -> alert(e.Message); None

    let initializeAssetWriter () =
        let filePath =
            Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments),
                "Temporary.mov")

        if File.Exists(filePath) then File.Delete(filePath)
        let videoUrl = NSUrl.FromFilename(filePath)
        let error : Ref<NSError> = ref null
        let writer = new AVAssetWriter(videoUrl, !> AVFileType.QuickTimeMovie, error)
        if error.Value <> null then Choice2Of2(error.Value.LocalizedDescription)
        else Choice1Of2(videoUrl, writer)

    let initializeSession f =
        //Create the capture session
        let session = new AVCaptureSession(SessionPreset = AVCaptureSession.PresetMedium)

        //Setup the video capture
        match AVCaptureDevice.DefaultDeviceWithMediaType(!> AVMediaType.Video) with
        | captureDevice when captureDevice <> null ->
            let input = AVCaptureDeviceInput.FromDevice(captureDevice)
            if input = null then
                Choice2Of2("No input - this won't work on the simulator, try a physical device")
            else
                session.AddInput(input)

                // create a VideoDataOutput and add it to the sesion
                let output = new AVCaptureVideoDataOutput(VideoSettings = new AVVideoSettings(CVPixelFormatType.CV32BGRA))

                // configure the output
                let queue = new MonoTouch.CoreFoundation.DispatchQueue("myQueue")
                output.SetSampleBufferDelegate(f, queue)
                session.AddOutput(output)
                Choice1Of2(session)
        | _ -> Choice2Of2("No captureDevice - this won't work on the simulator, try a physical device")

    let startRecording
        (initSession : unit -> AVCaptureSession option) 
        (initWriter : unit -> (NSUrl * AVAssetWriter) option)
        (initInputWriter : unit -> AVAssetWriterInput option) =
        try
            match initSession() with
            | None -> Choice2Of2("Couldn't initialize session")
            | Some(session : AVCaptureSession) ->
                match initWriter() with
                | None -> Choice2Of2("Couldn't initialize writer")
                | Some(url, writer : AVAssetWriter) ->
                    match initInputWriter() with
                    | None -> Choice2Of2("Couldn't initialize input writer")
                    | Some(inputWriter) ->
                        if not (writer.CanAddInput(inputWriter)) then
                            Choice2Of2("Couldn't add input writer to writer")
                        else
                            writer.AddInput(inputWriter)
                            session.StartRunning()
                            Choice1Of2(
                                {
                                    VideoUrl = url
                                    Session = session
                                    Writer = writer
                                    InputWriter = inputWriter
                                })
        with
            | e -> Choice2Of2(e.Message)

    let stopRecording onComplete = function 
        | {Session = session; Writer = writer; InputWriter = _} ->
            session.StopRunning()
            writer.FinishWriting(new NSAction(onComplete))

type LabelledView = {Label : UILabel; View : UIImageView}

type VideoCapture(labelledView) = 
    inherit AVCaptureVideoDataOutputSampleBufferDelegate()
    let frame = ref 0
    member val recording : Recording option = None with get, set

    member x.StartRecording () =
        match startRecording (x.InitializeSession) (x.InitializeAssetWriter) initializeInputWriter with
        | Choice1Of2(r) -> x.recording <- Some(r); true
        | Choice2Of2(m) -> alert(m); x.recording <- None; false

    member x.StopRecording () =
        match x.recording with
        | Some(r) -> try stopRecording (fun () -> x.MoveFinishedMovieToAlbum()) r with | e -> alert(e.Message)
        | None -> ()

    member x.InitializeSession () =
        match initializeSession x with
        | Choice1Of2(s) -> Some(s)
        | Choice2Of2(m) -> alert(m); None

    member x.InitializeAssetWriter () =
        match initializeAssetWriter() with
        | Choice1Of2(u, w) -> Some(u, w)
        | Choice2Of2(m) -> alert(m); None

    member x.MoveFinishedMovieToAlbum () =
        match x.recording, labelledView with
        | Some({VideoUrl = url; Session = _; Writer = _; InputWriter = _}), {Label = l; View = _} ->
            (new ALAssetsLibrary()).WriteVideoToSavedPhotosAlbum(
                url,
                new ALAssetsLibraryWriteCompletionDelegate(fun _ _ ->
                    l.BeginInvokeOnMainThread(fun () ->
                        l.Text <- "Movie saved to Album.")))
        | None, _ -> ()

    member x.DidOutputSampleBuffer (captureOutput, sampleBuffer : CMSampleBuffer, connection) =
        try
            try
                let lastSampleTime = sampleBuffer.PresentationTimeStamp
                match !frame, x.recording with
                | 0, Some({Session = _; Writer = w; InputWriter = _}) ->
                    w.StartWriting() |> ignore
                    w.StartSessionAtSourceTime(lastSampleTime)
                    frame := 1

                | _, Some({Session = _; Writer = _; InputWriter = iw}) ->
                    match labelledView with
                    | {Label = l; View = v} ->
                        v.BeginInvokeOnMainThread(fun () -> v.Image <- imageFromSampleBuffer(sampleBuffer))

                        l.BeginInvokeOnMainThread(fun () ->
                            let infoString =
                                if iw.ReadyForMoreMediaData then
                                    if not (iw.AppendSampleBuffer(sampleBuffer)) then
                                        "Failed to append sample buffer"
                                    else
                                        frame := !frame + 1
                                        String.Format("{0} frames captured", !frame)
                                else
                                    "Writer not ready"

                            l.Text <- infoString)
                | _ -> ()
            with
                | e -> alert(e.Message)
        finally
            sampleBuffer.Dispose()

type ContentView(fillColor, recordToggle) as x =
    inherit UIView()

    let lv =
        let bds = UIScreen.MainScreen.Bounds
        let imageBounds = new RectangleF (10.0f, 10.0f, bds.Width - 20.0f, bds.Height - 120.0f)
        {
            Label = new UILabel (new RectangleF (bds.Width - 150.0f, 10.0f, 140.0f, 50.0f))
            View = new UIImageView (imageBounds, BackgroundColor = UIColor.Blue)
        }

    do
        x.BackgroundColor <- fillColor
        ([lv.View; lv.Label; makeToggleButton recordToggle] : UIView list)
        |> List.iter (fun v -> x.AddSubview v)

    member val LabelledView = lv

type VideoCapturing =
    {Capture : VideoCapture option; IsRecording : bool}
    member x.Toggle lv =
        match x with
        | {Capture = Some(c); IsRecording = true} ->
            c.StopRecording()
            {x with IsRecording = false}
        | {Capture = _; IsRecording = false} ->
            let capture = new VideoCapture(lv())
            {Capture = Some(capture); IsRecording = capture.StartRecording()}
        | {Capture = None; IsRecording = true} ->
            failwith "No capture but recording. How could that happen?"

type VideoCaptureController(viewColor, title) =
    inherit UIViewController()
    member val recordingCapture = {Capture = None; IsRecording = false} with get, set

    override x.ViewDidLoad() =
        base.ViewDidLoad()
        x.Title <- title
        x.View <- new ContentView(viewColor, fun uib -> x.RecordToggle(fun s -> uib.SetTitle(s, UIControlState.Normal)))

    member x.RecordToggle (titleSetter : string -> unit) =
        let getLabelledView =
            match base.View with
            | :? ContentView as cv -> fun () -> cv.LabelledView
            | _ -> failwith "Base class is not a ContentView"

        x.recordingCapture <- x.recordingCapture.Toggle getLabelledView
        titleSetter <| if x.recordingCapture.IsRecording then "Stop" else "Record"

[<Register("AppDelegate")>]
type AppDelegate() =
    inherit UIApplicationDelegate()
    override __.FinishedLaunching(app, options) =
        let viewController = new VideoCaptureController(UIColor.Red, "Main")
        let window = new UIWindow(UIScreen.MainScreen.Bounds, RootViewController = viewController)
        window.MakeKeyAndVisible()
        true

type Application() =
    static member Main(args) = UIApplication.Main(args, null, "AppDelegate")

File: VideoCapture/Capture.cs
using System;
using System.Collections.Generic;
using System.Linq;

using MonoTouch.Foundation;
using MonoTouch.UIKit;
using System.Drawing;
using MonoTouch.CoreGraphics;
using MonoTouch.CoreAnimation;
using MonoTouch.AVFoundation;
using MonoTouch.CoreVideo;
using MonoTouch.CoreMedia;
using System.IO;
using System.Threading;
using MonoTouch.AssetsLibrary;

namespace VideoCapture
{
    public class VideoCapture : AVCaptureVideoDataOutputSampleBufferDelegate
    {
        public UIImageView ImageView { get; protected set; }

        public UILabel InfoLabel { get; protected set; }

        AVCaptureSession session;
        AVAssetWriter writer;
        AVAssetWriterInput inputWriter;
        CMTime lastSampleTime;
        NSUrl videoUrl;

        public VideoCapture(UIImageView imgView, UILabel label)
        {
            this.ImageView = imgView;
            this.InfoLabel = label;
        }

        public bool StartRecording()
        {
            try
            {
                session = MaybeInitializeSession();
                if (session == null)
                {
                    Failure.Alert("Couldn't initialize session");
                    return false;
                }
                writer = MaybeInitializeAssetWriter();
                if (writer == null)
                {
                    Failure.Alert("Couldn't initialize writer");
                    return false;
                }
                inputWriter = MaybeInitializeInputWriter();
                if (inputWriter == null)
                {
                    Failure.Alert("Couldn't initialize input writer");
                    return false;
                }
                if (!writer.CanAddInput(inputWriter))
                {
                    Failure.Alert("Couldn't add input writer to writer");
                    return false;
                }
                writer.AddInput(inputWriter);

                session.StartRunning();
                return true;
            }
            catch (Exception x)
            {
                Failure.Alert(x.Message);
                return false;
            }
        }

        public void StopRecording()
        {
            try
            {
                session.StopRunning();
                writer.FinishWriting(() => MoveFinishedMovieToAlbum());
            }
            catch (Exception x)
            {
                Failure.Alert(x.Message);
            }
        }

        //Protected
        protected AVCaptureSession MaybeInitializeSession()
        {
            //Create the capture session
            var session = new AVCaptureSession()
            {
                SessionPreset = AVCaptureSession.PresetMedium
            };

            //Setup the video capture
            var captureDevice = AVCaptureDevice.DefaultDeviceWithMediaType(AVMediaType.Video);
            if (captureDevice == null)
            {
                Failure.Alert("No captureDevice - this won't work on the simulator, try a physical device");
                return null;
            }
            var input = AVCaptureDeviceInput.FromDevice(captureDevice);
            if (input == null)
            {
                Failure.Alert("No input - this won't work on the simulator, try a physical device");
                return null;
            }
            session.AddInput(input);

            // create a VideoDataOutput and add it to the sesion
            var output = new AVCaptureVideoDataOutput()
            {
                VideoSettings = new AVVideoSettings(CVPixelFormatType.CV32BGRA),
            };

            // configure the output
            var queue = new MonoTouch.CoreFoundation.DispatchQueue("myQueue");
            output.SetSampleBufferDelegate(this, queue);
            session.AddOutput(output);

            return session;
        }

        protected AVAssetWriter MaybeInitializeAssetWriter()
        {
            var filePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), "Temporary.mov");
            if (File.Exists(filePath))
            {
                File.Delete(filePath);
            }
            videoUrl = NSUrl.FromFilename(filePath);

            NSError error;
            var writer = new AVAssetWriter(videoUrl, AVFileType.QuickTimeMovie, out error);
            if (error != null)
            {
                Failure.Alert(error.LocalizedDescription);
                return null;
            }
            return writer;
        }

        protected AVAssetWriterInput MaybeInitializeInputWriter()
        {
            try
            {
                var dictionary = NSDictionary.FromObjectsAndKeys(
                new NSObject[] { AVVideo.CodecH264, new NSNumber(640), new NSNumber(480) },
                new NSObject[] { AVVideo.CodecKey, AVVideo.WidthKey, AVVideo.HeightKey }
                );

                var writerInput = new AVAssetWriterInput(AVMediaType.Video, dictionary);
                writerInput.ExpectsMediaDataInRealTime = true;
                return writerInput;
            }
            catch (Exception x)
            {
                Failure.Alert(x.Message);
                return null;
            }
        }

        protected void MoveFinishedMovieToAlbum()
        {
            var lib = new ALAssetsLibrary();
            lib.WriteVideoToSavedPhotosAlbum(videoUrl, (o, s) => InfoLabel.BeginInvokeOnMainThread(() => InfoLabel.Text = "Movie saved to Album."));
        }

        //Implement AVCaptureVideoDataOutputSampleBufferDelegate
        int frame = 0;
        public override void DidOutputSampleBuffer(AVCaptureOutput captureOutput, CMSampleBuffer sampleBuffer, AVCaptureConnection connection)
        {
            try
            {
                lastSampleTime = sampleBuffer.PresentationTimeStamp;

                var image = ImageFromSampleBuffer(sampleBuffer);

                if (frame == 0)
                {
                    writer.StartWriting();
                    writer.StartSessionAtSourceTime(lastSampleTime);
                    frame = 1;
                }
                String infoString = "";
                if (inputWriter.ReadyForMoreMediaData)
                {
                    if (!inputWriter.AppendSampleBuffer(sampleBuffer))
                    {
                        infoString = "Failed to append sample buffer";
                    }
                    else
                    {
                        infoString = String.Format("{0} frames captured", frame++);
                    }
                }
                else
                {
                    infoString = "Writer not ready";
                }

                ImageView.BeginInvokeOnMainThread(() => ImageView.Image = image);
                InfoLabel.BeginInvokeOnMainThread(() => InfoLabel.Text = infoString);
            }
            catch (Exception e)
            {
                Failure.Alert(e.Message);
            }
            finally
            {
                sampleBuffer.Dispose();
            }
        }

        UIImage ImageFromSampleBuffer(CMSampleBuffer sampleBuffer)
        {
            // Get the CoreVideo image
            using (var pixelBuffer = sampleBuffer.GetImageBuffer() as CVPixelBuffer)
            {
                // Lock the base address
                pixelBuffer.Lock(0);
                // Get the number of bytes per row for the pixel buffer
                var baseAddress = pixelBuffer.BaseAddress;
                int bytesPerRow = pixelBuffer.BytesPerRow;
                int width = pixelBuffer.Width;
                int height = pixelBuffer.Height;
                var flags = CGBitmapFlags.PremultipliedFirst | CGBitmapFlags.ByteOrder32Little;
                // Create a CGImage on the RGB colorspace from the configured parameter above
                using (var cs = CGColorSpace.CreateDeviceRGB())
                using (var context = new CGBitmapContext(baseAddress, width, height, 8, bytesPerRow, cs, (CGImageAlphaInfo)flags))
                using (var cgImage = context.ToImage())
                {
                    pixelBuffer.Unlock(0);
                    return UIImage.FromImage(cgImage);
                }
            }
        }
    }

    public class Failure
    {
        public static void Alert(string msg)
        {
            new NSString().InvokeOnMainThread(() => new UIAlertView("Trouble", msg, null, "OK", null).Show());
        }
    }

    public class ContentView : UIView
    {
        public UIImageView ImageView { get; protected set; }

        public UILabel InfoLabel { get; protected set; }

        public ContentView(UIColor fillColor, EventHandler recordToggle)
        {
            BackgroundColor = fillColor;

            ImageView = new UIImageView(new RectangleF(10, 10, UIScreen.MainScreen.Bounds.Width - 20, UIScreen.MainScreen.Bounds.Height - 120));
            ImageView.BackgroundColor = UIColor.Blue;

            InfoLabel = new UILabel(new RectangleF(UIScreen.MainScreen.Bounds.Width - 150, 10, 140, 50));

            var toggleButton = UIButton.FromType(UIButtonType.RoundedRect);
            toggleButton.SetTitle("Record", UIControlState.Normal);
            toggleButton.Frame = new RectangleF(
                new PointF(UIScreen.MainScreen.Bounds.Width / 2 - toggleButton.IntrinsicContentSize.Width / 2, UIScreen.MainScreen.Bounds.Height - toggleButton.IntrinsicContentSize.Height - 50), toggleButton.IntrinsicContentSize);

            toggleButton.TouchUpInside += recordToggle;

            AddSubview(ImageView);
            AddSubview(InfoLabel);
            AddSubview(toggleButton);
        }
    }

    public class VideoCaptureController : UIViewController
    {
        String title;
        UIColor color;
        protected bool recording = false;
        protected VideoCapture videoCapture;

        public VideoCaptureController(UIColor viewColor, String title)
            : base()
        {
            this.title = title;

            color = viewColor;
        }

        public override void DidReceiveMemoryWarning()
        {
            // Releases the view if it doesn't have a superview.
            base.DidReceiveMemoryWarning();
        }

        public override void ViewDidLoad()
        {
            base.ViewDidLoad();

            Title = title;

            var cv = new ContentView(color, RecordToggle);

            View = cv;
        }

        protected void RecordToggle(Object sender, EventArgs e)
        {
            if (!recording)
            {
                ContentView cv = View as ContentView;
                videoCapture = new VideoCapture(cv.ImageView, cv.InfoLabel);
                recording = videoCapture.StartRecording();
            }
            else
            {
                videoCapture.StopRecording();
                recording = false;
            }
            var newTitle = recording ? "Stop" : "Record";
            (sender as UIButton).SetTitle(newTitle, UIControlState.Normal);
        }
    }

    [Register("AppDelegate")]
    public class AppDelegate : UIApplicationDelegate
    {
        UIWindow window;

        public override bool FinishedLaunching(UIApplication app, NSDictionary options)
        {
            var viewController = new VideoCaptureController(UIColor.Red, "Main");

            window = new UIWindow(UIScreen.MainScreen.Bounds);
            window.RootViewController = viewController;

            window.MakeKeyAndVisible();

            return true;
        }

    }

    public class Application
    {
        static void Main(string[] args)
        {
            UIApplication.Main(args, null, "AppDelegate");
        }
    }
}